src/com/amazon/corretto/crypto/provider/EvpKemKey.java

```
@@ -0,0 +1,42 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.corretto.crypto.provider;
+
+public abstract class EvpKemKey extends EvpKey {
+
+  private volatile Integer parameterSet;
+  private static final long serialVersionUID = 1;
+
+  private static native int getParameterSet(long ptr);
+
+  EvpKemKey(final InternalKey key, final boolean isPublicKey) {
+    super(key, EvpKeyType.KEM, isPublicKey);
+  }
+
+  public int getParameterSet() {
+    Integer result = parameterSet;
+    if (result == null) {
+      synchronized (this) {
+        result = parameterSet;
+        if (result == null) {
+          try {
+            Class<?> kemUtilsClass = Class.forName("com.amazon.corretto.crypto.provider.KemUtils");
+            java.lang.reflect.Method method =
+                kemUtilsClass.getDeclaredMethod("nativeGetParameterSet", long.class);
+            method.setAccessible(true);
+            result = use(ptr -> (Integer) method.invoke(null, ptr));
+          } catch (Exception e) {
+            result = -1;
+          }
+          parameterSet = result;
+        }
+      }
+    }
```

1. This code can probably be called once in the constructor and cached, instead of called at runtime. This would avoid the need for the extra thread synchronization code since only 1 thread can call the constructor.
2. You should probably throw an exception from the constructor if the pointer inside InternalKey is invalid and doesn't map to a valid MLKEM parameter size. 

As it is now, I think if you pass in an invalid pointer `getParameterSet()` might return `-1`, and as a result `getAlgorithm()` might return the string `ML-KEM--1`

---

Changed the check to be called in the constructor and stored the result.